ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/blobs/BlockBlobSidecarsTracker.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright ConsenSys Software Inc., 2023
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.blobs;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.deneb.BlobSidecar;
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.blocks.SlotAndBlockRoot;
+import tech.pegasys.teku.spec.datastructures.blocks.blockbody.versions.deneb.BeaconBlockBodyDeneb;
+import tech.pegasys.teku.spec.datastructures.networking.libp2p.rpc.BlobIdentifier;
+
+public class BlockBlobSidecarsTracker {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final SlotAndBlockRoot slotAndBlockRoot;
+
+  private final AtomicReference<Optional<BeaconBlockBodyDeneb>> blockBody =
+      new AtomicReference<>(Optional.empty());
+
+  private final Map<UInt64, BlobSidecar> blobSidecars = new ConcurrentHashMap<>();
+  private final SafeFuture<Void> blobSidecarsComplete = new SafeFuture<>();
+
+  public BlockBlobSidecarsTracker(final SlotAndBlockRoot slotAndBlockRoot) {
+    this.slotAndBlockRoot = slotAndBlockRoot;
+  }
+
+  public Map<UInt64, BlobSidecar> getBlobSidecars() {
+    return blobSidecars;
+  }
+
+  public SafeFuture<Void> getCompletionFuture() {
+    return blobSidecarsComplete;
+  }
+
+  public Optional<BeaconBlockBodyDeneb> getBlockBody() {
+    return blockBody.get();
+  }
+
+  public boolean containsBlobSidecar(final BlobIdentifier blobIdentifier) {
+    return Optional.ofNullable(blobSidecars.get(blobIdentifier.getIndex()))
+        .map(blobSidecar -> blobSidecar.getBlockRoot().equals(blobIdentifier.getBlockRoot()))
+        .orElse(false);
+  }
+
+  public Stream<BlobIdentifier> getMissingBlobSidecars() {
+    blockBody.get();
+    if (blockBody.get().isEmpty()) {
+      // TODO: block is still unknown.
+      //  Should we return all potential maxBlobsPerBlock BlobIdentifiers?
```

Comment:
ahh.. yeah...

Reply:
@zilm13 implemented

Reply:
implemented!

Reply:
even if, thinking about it again, I'm not sure it adds any value. If the objective is to try our best to import the block in time, this is not enough. If the block fetch is successful we may realize that we don't have requested all we need, so we will not be able to complete block import.

Reply:
Assume we fill only one blob once in a 2 blocks in the beginning in Mainnet. If every Teku requests all possible variants in advance, it will look very spammy. My proposal will not work all the times but at least it will not spam network

Reply:
true, but I don't like the probabilistic approach: if you're lucky that you missed the "left side" you can successfully recover, otherwise you cannot. So, for know I added the full "spammy" approach, hoping that we can tune constants in a way that we fall in that situation sufficiently rarely that we can afford to be "spammy". We will see in real devnets\testnets if worth to have it.